### PR TITLE
Add support for alternate boost-context backends

### PR DESCRIPTION
**Pull request by elfprince13**
The fcontext backend is the default high-performance backend.
The ucontext backend is needed when using Boost context in conjunction with ASAN.
The WinFibers backend...also exists.

cf. https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/cc/implementations__fcontext_t__ucontext_t_and_winfiber.html

### CODE REVIEW

**Review by tldahlgren — APPROVED** [review]
LGTM.  Deferring merge decision to maintainers/package experts.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/boost/package.py]: Please correct style issues:

```suggestion
            values=('fcontext', 'ucontext', 'winfib'),
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Please correct the style error; otherwise, LGTM.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [inline var/spack/repos/builtin/packages/boost/package.py:164]: What version of Boost introduced this feature? I'm working to make sure variants are correctly mapped to versions where they are actually available.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [inline var/spack/repos/builtin/packages/boost/package.py:494]: Assuming b2 won't raise any fuss if we pass `context-impl=fcontext`, there's no need to check for it here. That will remove this maintenance hazard in keeping the check here and the values of the variant in synch.

```suggestion
if '+context' in spec:
  context_impl = spec.variants['context-impl'].value
  options.append('context-impl=%s' % context_impl)
```

**Review by hainest — CHANGES_REQUESTED** [review]
It would be great if we could get a list of the properties that b2 supports, but I couldn't find a listing in the [docs](https://www.boost.org/doc/libs/1_79_0/tools/build/doc/html/index.html#bbv2.reference.features). How did you find this one?

File: var/spack/repos/builtin/packages/boost/package.py
Comment (elfprince13) [reply var/spack/repos/builtin/packages/boost/package.py:164]: The relevant pages are here:
https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/stack/sanitizers.html and here https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/cc/implementations__fcontext_t__ucontext_t_and_winfiber.html

It's available at least as far back as 1.65, haven't gone further back.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (elfprince13) [reply var/spack/repos/builtin/packages/boost/package.py:494]: The documentation page I linked above only discusses the possibility of `ucontext` or `winfib` as arguments.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [reply var/spack/repos/builtin/packages/boost/package.py:164]: I'll run my automated tests and check.  **edit**: We don't need to hold this PR for this.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (elfprince13) [reply var/spack/repos/builtin/packages/boost/package.py]: Sorted in [65d7b34](https://github.com/spack/spack/pull/30496/commits/65d7b34ce1839e576f377bba64a5eb9c256a2a58)

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [reply var/spack/repos/builtin/packages/boost/package.py:494]: In 1.79.0,

``` console
$ ./bootstrap.sh --with-libraries=context

$ ./b2 context-impl=moo
/boost_1_79_0/tools/build/src/build/feature.jam:491: in validate-value-string from module feature
error: "moo" is not a known value of feature <context-impl>
error: legal values: "fcontext" "ucontext" "winfib"

$ ./b2 context-impl=fcontext
  ...
    - context                  : building
  ...
The Boost C++ Libraries were successfully built!
```

Let's assume for now that always works. I'll add it to my TODO list to double-check.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (elfprince13) [reply var/spack/repos/builtin/packages/boost/package.py:164]: Incidentally there's an option for supporting valgrind that would be great to get rolled in as well (as a separate PR):

https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/stack/valgrind.html

**Review by tldahlgren — DISMISSED** [review]
Removing my blocking review related to style check.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [reply var/spack/repos/builtin/packages/boost/package.py:164]: Oh, very nice! I will try to see about getting at least a reference list of the features/properties.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (elfprince13) [reply var/spack/repos/builtin/packages/boost/package.py:494]: Cool - i just verified `fcontext` is present as an option at least as far back as 1.65:
https://github.com/boostorg/context/blob/boost-1.65.0/build/Jamfile.v2

I'll go ahead and remove that extra check.

I will note that the special-casing for the defines introduced in [65d7b34](https://github.com/spack/spack/pull/30496/commits/65d7b34ce1839e576f377bba64a5eb9c256a2a58) should be kept as `BOOST_USE_FCONTEXT` is not used anywhere.

Comment (hainest) [reply var/spack/repos/builtin/packages/boost/package.py:494]: Oh, that's... neat. Particularly since ucontext was [removed](https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/rationale/other_apis_.html#ucontext) in POSIX.1-2008. Boost, why did you do that?!

Out of curiosity, which backend are you using?

Comment (elfprince13) [reply var/spack/repos/builtin/packages/boost/package.py:494]: fcontext is definitely the sensible default for most use-cases, but ucontext is the only way to cooperate with ASAN if you're doing memory debugging (or was - it seems to have broken at some point recently since ASAN promptly reported a bug in boost itself when I switched)

Comment (hainest) [reply var/spack/repos/builtin/packages/boost/package.py:494]: Oh sorry, yes, you mentioned that in your commit. It's interesting ASAN is relying on the deprecated/removed interface. I would assume it's really difficult to instrument the fcontext usage.

**Comment by hainest** [comment]
Looks like the pipeline is stuck.

@spackbot re-run pipeline

**Review by hainest — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

**Pull request merged**
